client/allocrunner/taskrunner/plugin_supervisor_hook.go

```diff
@@ -0,0 +1,272 @@
+
+	hook := &csiPluginSupervisorHook{
+		alloc:            runner.Alloc(),
+		runner:           runner,
+		logger:           logger,
+		task:             task,
+		mountPoint:       pluginRoot,
+		shutdownCtx:      shutdownCtx,
+		shutdownCancelFn: cancelFn,
+		eventEmitter:     eventEmitter,
+	}
+
+	return hook
+}
+
+func (*csiPluginSupervisorHook) Name() string {
+	return "csi_plugin_supervisor"
+}
+
+// Prestart is called before the task is started including after every
+// restart. This requires that the mount paths for a plugin be idempotent,
+// despite us not knowing the name of the plugin ahead of time.
+// Because of this, we use the allocid_taskname as the unique identifier for a
+// plugin on the filesystem.
+func (h *csiPluginSupervisorHook) Prestart(ctx context.Context,
+	req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	// TODO: Figure out correct mode
+	err := os.MkdirAll(h.mountPoint, 0750)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("failed to create mount point: %v", err)
+	}
+
+	configMount := &drivers.MountConfig{
+		TaskPath:        h.task.CSIPluginConfig.PluginMountDir,
+		HostPath:        h.mountPoint,
+		Readonly:        false,
+		PropagationMode: "bidirectional",
+	}
+
+	mounts := ensureMountpointInserted(h.runner.hookResources.getMounts(), configMount)
+	h.runner.hookResources.setMounts(mounts)
+
+	resp.Done = true
+	return nil
+}
+
+// Poststart is called after the task has started. Poststart is not
+// called if the allocation is terminal.
+//
+// The context is cancelled if the task is killed.
+func (h *csiPluginSupervisorHook) Poststart(_ context.Context, _ *interfaces.TaskPoststartRequest, _ *interfaces.TaskPoststartResponse) error {
+	// If we're already running the supervisor routine, then we don't need to try
+	// and restart it here as it only terminates on `Stop` hooks.
+	h.runningLock.Lock()
+	if h.running == true {
+		h.runningLock.Unlock()
+		return nil
+	}
+	h.runningLock.Unlock()
+
+	go h.ensureSupervisorLoop(h.shutdownCtx)
+	return nil
+}
+
+// ensureSupervisorLoop starts a goroutine that will runs on an interval until
+// the passed context is cancelled.
+//
+// The supervisor works by:
+// - Initially waiting for the plugin to become available. This loop is expensive
+//   and may do things like create new gRPC Clients on every iteration.
+// - After recieving an initial healthy status, it will inform the plugin catalog
+//   of the plugin, registering it with the plugins fingerprinted capabilities.
+// - We then perform a more lightweight check, simply probing the plugin on a less
+//   frequent interval to ensure it is still alive, emiting task events when this
+//   status changes.
+//
+// Deeper fingerprinting of the plugin is not yet implemented, but may happen here
+// or in the plugin catalog in the future.
+func (h *csiPluginSupervisorHook) ensureSupervisorLoop(ctx context.Context) {
```

Comment:
We don't implement an `Update` method for this hook that could run concurrently with `PostStart`, which is (I think?) why we usually see these locks in task runner hooks. Is there a safe way to implement `Update` or do we assume we'll have to blow away the whole task if we want to change CSI plugin config?

Reply:
We'd have to blow away the whole task because it would change mount points, unfortunately :(

Reply:
(In some cases, that would also disrupt _other_ running tasks, depending on the configuration change)